DOC/TST: Update the parquet (pyarrow >= 0.15) docs and tests regarding Categorical support

galuhsahid opened the pull request:
- [x] closes #27955
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

galuhsahid committed:
TST: Update pyarrow tests to test parquet faithful roundtrip for Categorical support

Fixes #27955

galuhsahid committed:
DOC: Update parquet docs regarding Categorical support

Fixes #27955

galuhsahid committed:
DOC: Add pyarrow test to test parquet faithful roundtrip for Categorical support

Fixes #27955

galuhsahid commented:
Several questions/thoughts:
1. Should I also add a test in this part: https://github.com/pandas-dev/pandas/blob/802f67046bbae0a815b2fe9d20d2217485bbc942/pandas/tests/io/test_parquet.py#L409 as well? I thought adding a test there would be a duplicate of what I'm doing in https://github.com/pandas-dev/pandas/blob/802f67046bbae0a815b2fe9d20d2217485bbc942/pandas/tests/io/test_parquet.py#L451, but maybe I'm missing something
2. In https://github.com/apache/arrow/pull/5110 the test in pyarrow demonstrates how category values will be preserved regardless whether they occur in the data or not; should the test in pandas demonstrate this as well?

jorisvandenbossche reviewed (commented):
Thanks a lot for the PR! 

Looks very good. One suggestion I have is to add a categorical column in the first example in the docs (the `df` with columns of all different types that gets created a bit below the place you edited in io.rst).

jorisvandenbossche commented:
> Should I also add a test in this part: 

I think your analysis is correct that it would be duplicative. 
I think once we don't have the version check anymore (in the future if we would only support pyarrow >= 0.15), then we can move the ``test_categorical`` into the `test_basic`. But for now I would keep it as is.

> In apache/arrow#5110 the test in pyarrow demonstrates how category values will be preserved regardless whether they occur in the data or not; should the test in pandas demonstrate this as well?

Yes, I think that is nice to test. In addition, you could also test if the `ordered` flag is preserved as well. You can add additional columns to the `df` with different cases in `test_categorical`.

galuhsahid committed:
DOC: Add categorical column in example DataFrame

Closes #27955

galuhsahid committed:
TST: Add test for null, out-of-order values, and unobserved category

Closes #27955

galuhsahid commented:
Thanks for the review! I've added another test for null, out-of-order values & unobserved category and updated the doc as well.

However when I tried to add another test to check if order is preserved, it seems like it's not preserved. I'm not really sure if this is expected or if I'm missing something:

```
import pandas as pd
from pandas.io.parquet import read_parquet, to_parquet

df = pd.DataFrame()
df["a"] = pd.Categorical(["a", "b", "c", "a"], categories=["b", "c", "d"], ordered=True)

df.to_parquet(<path>)
actual = read_parquet(<path>)

df["a"]    
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b < c < d]

actual["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b, c, d]

```

If it's expected then maybe I can add another test that expects it to still return a categorical data type without order, if needed.

jorisvandenbossche commented:
> However when I tried to add another test to check if order is preserved, it seems like it's not preserved. I'm not really sure if this is expected or if I'm missing something:

Indeed, I see the same trying your code. That is certainly not expected I think. Can you open an issue for this? (Arrow uses a JIRA tracker: https://issues.apache.org/jira/projects/ARROW/issues; I can also open an issue if you want)

jorisvandenbossche reviewed (approved):
The additional changes look good to me, thanks!

jorisvandenbossche commented:
Hmm, the failure on Azure is actually related:

```=================================== FAILURES ===================================
_____________________ TestParquetPyArrow.test_categorical ______________________
[gw0] darwin -- Python 3.5.6 /Users/vsts/miniconda3/envs/pandas-dev/bin/python

self = <pandas.tests.io.test_parquet.TestParquetPyArrow object at 0x146354a90>
pa = 'pyarrow'

    def test_categorical(self, pa):
    
        # supported in >= 0.7.0
        df = pd.DataFrame()
        df["a"] = pd.Categorical(list("abcdef"))
    
        # test for null, out-of-order values, and unobserved category
        dtype = pd.CategoricalDtype(["foo", "bar", "baz"])
        df["b"] = pd.Categorical.from_codes(codes=[1, 0, 0, 1, -1, 1], dtype=dtype)
    
        if LooseVersion(pyarrow.__version__) >= LooseVersion("0.15.0"):
            check_round_trip(df, pa)
        else:
            # de-serialized as object for pyarrow < 0.15
            expected = df.assign(a=df.a.astype(object))
>           check_round_trip(df, pa, expected=expected)

pandas/tests/io/test_parquet.py:468: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/tests/io/test_parquet.py:175: in check_round_trip
    compare(repeat)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

repeat = 2

    def compare(repeat):
        for _ in range(repeat):
            df.to_parquet(path, **write_kwargs)
            with catch_warnings(record=True):
                actual = read_parquet(path, **read_kwargs)
    
>           tm.assert_frame_equal(expected, actual, check_names=check_names)
E           AssertionError: Attributes are different
E           
E           Attribute "dtype" are different
E           [left]:  CategoricalDtype(categories=['foo', 'bar', 'baz'], ordered=False)
E           [right]: object

pandas/tests/io/test_parquet.py:171: AssertionError
```
It seems that the `expected` still has the categorical dtype (so the `assign` failed to overwrite the original column), but not sure why this only happens on the mac/windows builds ..

jorisvandenbossche commented:
Ah, no, I already see ;). In the `expected` for the branch for older pyarrow versions, you still need to convert column `'b'` to object dtype as well

galuhsahid committed:
TST: Convert column b in expected to object for pyarrow < 0.15

Closes #27955

galuhsahid commented:
> Indeed, I see the same trying your code. That is certainly not expected I think. Can you open an issue for this? (Arrow uses a JIRA tracker: https://issues.apache.org/jira/projects/ARROW/issues; I can also open an issue if you want)

I've opened the issue here: https://issues.apache.org/jira/browse/ARROW-6302

jorisvandenbossche commented:
Thanks!

galuhsahid committed:
Merge branch 'master' into parquet-categorical

* master: (40 commits)
  DOC: Fix GL01 and GL02 errors in the docstrings (#27988)
  Remove Encoding of values in char** For Labels (#27618)
  TYPING: more type hints for io.formats.printing (#27765)
  TST: fix compression tests when run without virtualenv/condaenv (#28051)
  DOC: Start 0.25.2 (#28111)
  DOC: Fix docstrings lack of punctuation (#28031)
  DOC: Remove alias for numpy.random.randn from the docs (#28082)
  DOC: update GroupBy.head()/tail() documentation (#27844)
  BUG: timedelta merge asof with tolerance (#27650)
  BUG: Series.rename raises error on values accepted by Series construc… (#27814)
  Preserve index when setting new column on empty dataframe. (#26471)
  BUG: Fixed groupby quantile for listlike q (#27827)
  BUG: iter with readonly values, closes #28055 (#28074)
  TST: non-strict xfail for period test (#28072)
  DOC: Update whatsnew (#28073)
  CI: disable codecov (#28065)
  CI: Set SHA for codecov upload (#28067)
  BUG: Correct the previous bug fixing on xlim for plotting (#28059)
  CI: Add pip dependence explicitly (#28008)
  DOC: Change document code prun in a row (#28029)
  ...

galuhsahid committed:
TST: Add test for ordered flag

Closes #27955

galuhsahid committed:
TST: Add more rows for ordered flag test

Closes #27955

galuhsahid commented:
@jorisvandenbossche I've added the test for the ordered flag. Is there any other test that you think we should have for now?

galuhsahid committed:
DOC: Change to_parquet to DataFrame.to_parquet

Closes #27955

galuhsahid committed:
TST: Replace from_codes

Closes #27955

galuhsahid committed:
TST: Simplify expected conversion to object

Closes #27955

galuhsahid commented on doc/source/user_guide/io.rst:
```diff
@@ -4702,7 +4702,7 @@ Several caveats.
   force including or omitting indexes with the ``index`` argument, regardless of the underlying engine.
 * Index level names, if specified, must be strings.
-* Categorical dtypes can be serialized to parquet, but will de-serialize as ``object`` dtype.
+* Categorical dtypes for non-string types can be serialized to parquet, but will de-serialize as ``object`` dtype.
```
- I've just tried categorical dtypes using integer and in pyarrow it gets de-serialized into float. Not sure if it's expected, if someone can confirm it'd be great.

```
>>> import pandas as pd
>>> from pandas.io.parquet import read_parquet, to_parquet

>>> df["a"] = pd.Categorical(["a", "b", "c", "a"], categories=["b", "c", "d"], ordered=True)
>>> df["b"] = pd.Categorical([1, 2, 3, 1], categories=[2, 3, 4], ordered=True)

>>> df["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b < c < d]

>>> df["b"]
0    NaN
1      2
2      3
3    NaN
Name: b, dtype: category
Categories (3, int64): [2 < 3 < 4]

>>> df.to_parquet("test_pyarrow.parquet", engine="pyarrow")
>>> actual_pyarrow = df.read_parquet("test_pyarrow.parquet", engine="fastparquet")

>>> actual_pyarrow["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b, c, d]

>>> actual_pyarrow["b"]
0    NaN
1    2.0
2    3.0
3    NaN
Name: b, dtype: float64
```

- With fastparquet, both string and non-string types get de-serialized back as category, however it does not preserve order for both, while in pyarrow (for string) we do preserve order. Should we make this clear in the docs? 
```
>>> import pandas as pd
>>> from pandas.io.parquet import read_parquet, to_parquet

>>> df["a"] = pd.Categorical(["a", "b", "c", "a"], categories=["b", "c", "d"], ordered=True)
>>> df["b"] = pd.Categorical([1, 2, 3, 1], categories=[2, 3, 4], ordered=True)

>>> df["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b < c < d]

>>> df["b"]
0    NaN
1      2
2      3
3    NaN
Name: b, dtype: category
Categories (3, int64): [2 < 3 < 4]

>>> df.to_parquet("test_fastparquet.parquet", engine="fastparquet")
>>> actual_fastparquet = df.read_parquet("test_fastparquet.parquet", engine="fastparquet")

>>> actual_fastparquet["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b, c, d]

>>> actual_fastparquet["b"]
0    NaN
1      2
2      3
3    NaN
Name: b, dtype: category
Categories (3, int64): [2, 3, 4]
```

jorisvandenbossche commented:
> I've just tried categorical dtypes using integer and in pyarrow it gets de-serialized into float. Not sure if it's expected, if someone can confirm it'd be great.

For Arrow, that is expected, see https://issues.apache.org/jira/browse/ARROW-6140 (the fact that it is float and not int is because there are missing values)

> With fastparquet, both string and non-string types get de-serialized back as category, however it does not preserve order for both, while in pyarrow (for string) we do preserve order. Should we make this clear in the docs?

That could be mentioned yes.

galuhsahid committed:
DOC: Update doc to clarify the differences between pyarrow and fastparquet

Closes #27955

galuhsahid commented:
I've updated the doc to explain more about the differences regarding how the categorical support works between pyarrow and fastparquet

galuhsahid committed:
DOC: Remove whitespace

TomAugspurger commented on pandas/tests/io/test_parquet.py:
```diff
@@ -451,11 +453,26 @@ def test_unsupported(self, pa):
-        # de-serialized as object
-        expected = df.assign(a=df.a.astype(object))
-        check_round_trip(df, pa, expected=expected)
+        if LooseVersion(pyarrow.__version__) >= LooseVersion("0.15.0"):
```
@jorisvandenbossche this isn't released yet, right? Should we wait to merge until 0.15.0 is released?

jorisvandenbossche replied:
Yes, this isn't released yet. I can assure that it runs locally for me on Arrow master (if I change the version check to `> LooseVersion("0.14.1.dev")`), so I am OK to merge this, but also fine to wait a bit more (0.15.0 will normally happen somewhere end of next week)

galuhsahid replied:
Since Arrow 0.15.0 has been released, can we merge this now?

galuhsahid committed:
Merge branch 'master' into parquet-categorical

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -174,6 +174,7 @@ Categorical
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
 - Bug in :meth:`Categorical.astype` where ``NaN`` values were handled incorrectly when casting to int (:issue:`28406`)
 - :meth:`Categorical.searchsorted` and :meth:`CategoricalIndex.searchsorted` now work on unordered categoricals also (:issue:`21667`)
+- Added test to assert roundtripping to parquet with :func:`DataFrame.to_parquet` or :func:`read_parquet` will preserve Categorical dtypes for string types (:issue:`27955`)
```
this note is not really needed, but ok

jreback commented on doc/source/user_guide/io.rst:
```diff
@@ -4734,7 +4735,8 @@ See the documentation for `pyarrow <https://arrow.apache.org/docs/python/>`__ an
                       'f': pd.date_range('20130101', periods=3),
-                      'g': pd.date_range('20130101', periods=3, tz='US/Eastern')})
+                      'g': pd.date_range('20130101', periods=3, tz='US/Eastern'),
+                      'h': pd.Categorical(list('abc'))})
```
can we add an ordered categorical as well?

jreback reviewed (changes requested):
pls merge master

galuhsahid committed:
Merge branch 'master' into parquet-categorical

* master: (22 commits)
  DOC: fix PR09,PR08 errors for pandas.Timestamp (#28739)
  WEB: Add diversity note to team.md (#28630)
  DOC: Minor fixes in pandas/testing.py docstring. (#28752)
  TST: port maybe_promote tests from #23982 (#28764)
  Bugfix/groupby datetime issue (#28569)
  reenable codecov (#28750)
  CLN: Centralised _check_percentile (#27584)
  DEPR: Deprecate Index.set_value (#28621)
  CLN: Fix typo in contributing.rst (#28761)
  Fixed docstring errors in pandas.period range and pandas.PeriodIndex (#28756)
  BUG: Fix TypeError raised in libreduction (#28643)
  DOC: Pandas.Series.drop docstring PR02 (#27976) (#28742)
  DOC: Fixed doctring errors PR08, PR09 in pandas.io (#28748)
  TST: Fix broken test cases where Timedelta/Timestamp raise (#28729)
  REF: Consolidate alignment calls in DataFrame ops (#28638)
  BUG: Fix dep generation (#28734)
  Added doctstring to fixture (#28727)
  DOC: Fixed PR06 docstrings errors in pandas.timedelta_range (#28719)
  replaced safe_import with a corresponding test decorator (#28731)
  BUG: Fix RangeIndex.get_indexer for decreasing RangeIndex (#28680)
  ...

galuhsahid committed:
DOC: Add ordered categorical

jorisvandenbossche commented:
@galuhsahid Thanks a lot for this PR! (and for the fixes on the pyarrow side as well! :-))
